merkle/hashers/tree_hasher.go

```diff
@@ -33,6 +35,32 @@ type LogHasher interface {
 	Size() int
 }
 
+// InplaceLogHasher provides the hash functions needed to compute dense merkle
+// trees.
+// Implementations may hold state and are not expected to be thread
+// safe. They should not be shared between multiple callers. Where performance
+// is not critical the LogHasher interface provides a safer alternative.
+//
+// Note: Implementations are not expected to be thread safe. The caller must
+// ensure that concurrent calls are not made and is responsible for making
+// copies of slices where necessary to avoid data being overwritten.
+type InplaceLogHasher interface {
+	// EmptyRootInto supports returning a special case for the root of an empty tree.
+	// The passed in slice is mutated. The caller is responsible for making copies
+	// where necessary. Not thread safe.
+	EmptyRootInto([]byte) []byte
+	// HashLeafInto computes the hash of a leaf into an existing slice,
+	// which is mutated. The leaf and res slices may be the same. Not thread safe.
```

Comment:
After a second look:

In `HashChildrenInto`, it feels natural to allow `res == l/r` as they are the same "hash" type, and hence all have `cap(...) >= Size()`.

It feels less natural to allow `res == leaf` in `HashLeafInto` though, as they represent different types of data. Moreover, it's more error-prone, as the caller must ensure `cap(leaf) >= h.Size()`, even if `len(leaf) < h.Size()` -- quite an unnatural thing to do, because whoever produces the `leaf` slice might be not aware of the hasher.

I suggest removing this sentence from `HashLeafInto`, to not encourage such use-case.

Reply:
It's not necessary that the capacity is large enough as the value is produced by truncate followed by append. Test added to demonstrate.

I'm going to leave the doc in as it does work though it might not be a great idea to do it.